### ISSUE TITLE
replication resource allowing children to vote twice

### ISSUE THREAD

**Issue by trel** [irods/irods#2884]
First posted at https://groups.google.com/forum/#!topic/irod-chat/T3ayz4UzBl8:

> While the function of round robin is executed well with 3 elements it fails to perform with 2. It always chooses the same one.
> We have moved resources around and the behavior is persistent. 

The simple case works as expected, the replicas alternate between the two children:

```
$ ilsresc
demoResc
rr:roundrobin
├── a
└── b

$ iput -R rr foo foo1
$ ils -L
/tempZone/home/rods:
  rods              0 rr;b          423 2015-09-17.09:01 & foo1
        generic    /tmp/b/home/rods/foo1
$ iput -R rr foo foo2
$ ils -L
/tempZone/home/rods:
  rods              0 rr;b          423 2015-09-17.09:01 & foo1
        generic    /tmp/b/home/rods/foo1
  rods              0 rr;a          423 2015-09-17.09:01 & foo2
        generic    /tmp/a/home/rods/foo2
$ iput -R rr foo foo3
$ ils -L
/tempZone/home/rods:
  rods              0 rr;b          423 2015-09-17.09:01 & foo1
        generic    /tmp/b/home/rods/foo1
  rods              0 rr;a          423 2015-09-17.09:01 & foo2
        generic    /tmp/a/home/rods/foo2
  rods              0 rr;b          423 2015-09-17.09:01 & foo3
        generic    /tmp/b/home/rods/foo3
$ iput -R rr foo foo4
$ ils -L
/tempZone/home/rods:
  rods              0 rr;b          423 2015-09-17.09:01 & foo1
        generic    /tmp/b/home/rods/foo1
  rods              0 rr;a          423 2015-09-17.09:01 & foo2
        generic    /tmp/a/home/rods/foo2
  rods              0 rr;b          423 2015-09-17.09:01 & foo3
        generic    /tmp/b/home/rods/foo3
  rods              0 rr;a          423 2015-09-17.09:03 & foo4
        generic    /tmp/a/home/rods/foo4
$ iput -R rr foo foo5
$ ils -L
/tempZone/home/rods:
  rods              0 rr;b          423 2015-09-17.09:01 & foo1
        generic    /tmp/b/home/rods/foo1
  rods              0 rr;a          423 2015-09-17.09:01 & foo2
        generic    /tmp/a/home/rods/foo2
  rods              0 rr;b          423 2015-09-17.09:01 & foo3
        generic    /tmp/b/home/rods/foo3
  rods              0 rr;a          423 2015-09-17.09:03 & foo4
        generic    /tmp/a/home/rods/foo4
  rods              0 rr;b          423 2015-09-17.09:03 & foo5
        generic    /tmp/b/home/rods/foo5
```

The more complicated structure from the original report exhibits the broken behavior:

```
$ ilsresc
demoResc
pt:passthru
└── repl:replication
    ├── rr2:roundrobin
    │   ├── a
    │   └── b
    └── rr3:roundrobin
        ├── x
        ├── y
        └── z
```

The roundrobin with two children *does* only place replicas on resource 'a'.

```
$ iput -R pt foo foo1
$ ils -L
/tempZone/home/rods:
  rods              0 pt;repl;rr3;y          423 2015-09-17.09:10 & foo1
        generic    /tmp/y/home/rods/foo1
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:10 & foo1
        generic    /tmp/a/home/rods/foo1
$ iput -R pt foo foo2
$ ils -L
/tempZone/home/rods:
  rods              0 pt;repl;rr3;y          423 2015-09-17.09:10 & foo1
        generic    /tmp/y/home/rods/foo1
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:10 & foo1
        generic    /tmp/a/home/rods/foo1
  rods              0 pt;repl;rr3;z          423 2015-09-17.09:12 & foo2
        generic    /tmp/z/home/rods/foo2
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:12 & foo2
        generic    /tmp/a/home/rods/foo2
$ iput -R pt foo foo3
$ ils -L
/tempZone/home/rods:
  rods              0 pt;repl;rr3;y          423 2015-09-17.09:10 & foo1
        generic    /tmp/y/home/rods/foo1
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:10 & foo1
        generic    /tmp/a/home/rods/foo1
  rods              0 pt;repl;rr3;z          423 2015-09-17.09:12 & foo2
        generic    /tmp/z/home/rods/foo2
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:12 & foo2
        generic    /tmp/a/home/rods/foo2
  rods              0 pt;repl;rr3;x          423 2015-09-17.09:12 & foo3
        generic    /tmp/x/home/rods/foo3
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:12 & foo3
        generic    /tmp/a/home/rods/foo3
$ iput -R pt foo foo4
$ ils -L
/tempZone/home/rods:
  rods              0 pt;repl;rr3;y          423 2015-09-17.09:10 & foo1
        generic    /tmp/y/home/rods/foo1
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:10 & foo1
        generic    /tmp/a/home/rods/foo1
  rods              0 pt;repl;rr3;z          423 2015-09-17.09:12 & foo2
        generic    /tmp/z/home/rods/foo2
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:12 & foo2
        generic    /tmp/a/home/rods/foo2
  rods              0 pt;repl;rr3;x          423 2015-09-17.09:12 & foo3
        generic    /tmp/x/home/rods/foo3
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:12 & foo3
        generic    /tmp/a/home/rods/foo3
  rods              0 pt;repl;rr3;y          423 2015-09-17.09:12 & foo4
        generic    /tmp/y/home/rods/foo4
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:12 & foo4
        generic    /tmp/a/home/rods/foo4
$ iput -R pt foo foo5
$ ils -L
/tempZone/home/rods:
  rods              0 pt;repl;rr3;y          423 2015-09-17.09:10 & foo1
        generic    /tmp/y/home/rods/foo1
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:10 & foo1
        generic    /tmp/a/home/rods/foo1
  rods              0 pt;repl;rr3;z          423 2015-09-17.09:12 & foo2
        generic    /tmp/z/home/rods/foo2
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:12 & foo2
        generic    /tmp/a/home/rods/foo2
  rods              0 pt;repl;rr3;x          423 2015-09-17.09:12 & foo3
        generic    /tmp/x/home/rods/foo3
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:12 & foo3
        generic    /tmp/a/home/rods/foo3
  rods              0 pt;repl;rr3;y          423 2015-09-17.09:12 & foo4
        generic    /tmp/y/home/rods/foo4
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:12 & foo4
        generic    /tmp/a/home/rods/foo4
  rods              0 pt;repl;rr3;z          423 2015-09-17.09:12 & foo5
        generic    /tmp/z/home/rods/foo5
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:12 & foo5
        generic    /tmp/a/home/rods/foo5
```

**Comment by trel** [comment]
I think the problem is a shared context between multiple roundrobin counters...

With two roundrobins, with two and four children respectively, both roundrobins 'skip', or, count by two.  When there were only three children, this exhibited 'normal' behavior, since skipping one still appears to walk the list of three.

```
$ ilsresc
demoResc
pt:passthru
└── repl:replication
    ├── rr2:roundrobin
    │   ├── a
    │   └── b
    └── rr4:roundrobin
        ├── w
        ├── x
        ├── y
        └── z
```

Note replicas are only on 'x' and 'z', and 'a':

```
$ iput -R pt foo foo6
$ ils -L
/tempZone/home/rods:
  rods              0 pt;repl;rr4;x          423 2015-09-17.09:27 & foo6
        generic    /tmp/x/home/rods/foo6
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:27 & foo6
        generic    /tmp/a/home/rods/foo6
$ iput -R pt foo foo7
$ ils -L
/tempZone/home/rods:
  rods              0 pt;repl;rr4;x          423 2015-09-17.09:27 & foo6
        generic    /tmp/x/home/rods/foo6
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:27 & foo6
        generic    /tmp/a/home/rods/foo6
  rods              0 pt;repl;rr4;z          423 2015-09-17.09:27 & foo7
        generic    /tmp/z/home/rods/foo7
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:27 & foo7
        generic    /tmp/a/home/rods/foo7
$ iput -R pt foo foo8
$ ils -L
/tempZone/home/rods:
  rods              0 pt;repl;rr4;x          423 2015-09-17.09:27 & foo6
        generic    /tmp/x/home/rods/foo6
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:27 & foo6
        generic    /tmp/a/home/rods/foo6
  rods              0 pt;repl;rr4;z          423 2015-09-17.09:27 & foo7
        generic    /tmp/z/home/rods/foo7
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:27 & foo7
        generic    /tmp/a/home/rods/foo7
  rods              0 pt;repl;rr4;x          423 2015-09-17.09:27 & foo8
        generic    /tmp/x/home/rods/foo8
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:27 & foo8
        generic    /tmp/a/home/rods/foo8
$ iput -R pt foo foo9
$ ils -L
/tempZone/home/rods:
  rods              0 pt;repl;rr4;x          423 2015-09-17.09:27 & foo6
        generic    /tmp/x/home/rods/foo6
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:27 & foo6
        generic    /tmp/a/home/rods/foo6
  rods              0 pt;repl;rr4;z          423 2015-09-17.09:27 & foo7
        generic    /tmp/z/home/rods/foo7
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:27 & foo7
        generic    /tmp/a/home/rods/foo7
  rods              0 pt;repl;rr4;x          423 2015-09-17.09:27 & foo8
        generic    /tmp/x/home/rods/foo8
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:27 & foo8
        generic    /tmp/a/home/rods/foo8
  rods              0 pt;repl;rr4;z          423 2015-09-17.09:27 & foo9
        generic    /tmp/z/home/rods/foo9
  rods              1 pt;repl;rr2;a          423 2015-09-17.09:27 & foo9
        generic    /tmp/a/home/rods/foo9
```

**Comment by jasoncoposky** [comment]
To be clear, there is no shared context between nodes within the hierarchy.  The issue is that a fix in the replication node from #2789 requests the children to vote twice, once with a create/write operation and once with the intended operation.  This allows the replication node's children to provide information for a potential write after the initial operation is performed.  The side effect with the round robin resource is that it selects a new child on a vote for a create operation which causes it to iterate more than desired in this configuration.  That behavior will be moved to the file_modified operation where the context is saved in the database.  The lesson here is that voting within the hierarchy should be entirely idempotent and no state should change within a node during this operation.  The file_modified operation is the appropriate place to change state.
